src/Features/Core/Portable/Diagnostics/LiveDiagnosticUpdateArgsId.cs

```diff
@@ -2,29 +2,35 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Diagnostics
 {
     internal class LiveDiagnosticUpdateArgsId : AnalyzerUpdateArgsId
     {
-        private readonly string _analyzerPackageName;
+        private string? _buildTool;
 
         public readonly object ProjectOrDocumentId;
         public readonly AnalysisKind Kind;
 
-        public LiveDiagnosticUpdateArgsId(DiagnosticAnalyzer analyzer, object projectOrDocumentId, AnalysisKind kind, string analyzerPackageName)
+        public LiveDiagnosticUpdateArgsId(DiagnosticAnalyzer analyzer, object projectOrDocumentId, AnalysisKind kind)
             : base(analyzer)
         {
             Contract.ThrowIfNull(projectOrDocumentId);
 
             ProjectOrDocumentId = projectOrDocumentId;
             Kind = kind;
-
-            _analyzerPackageName = analyzerPackageName;
         }
 
-        public override string BuildTool => _analyzerPackageName;
+        public override string BuildTool => _buildTool ??= ComputeBuildTool();
+
+        private string ComputeBuildTool()
+        {
+            return Analyzer.IsBuiltInAnalyzer() || Analyzer == FileContentLoadAnalyzer.Instance || Analyzer == GeneratorDiagnosticsPlaceholderAnalyzer.Instance
```

Comment:
```suggestion
            return Analyzer.IsBuiltInAnalyzer()
```

`IsBuiltInAnalyzer()` already covers the other cases: 
https://github.com/dotnet/roslyn/blob/02dce0619b4155e1b469c61665c9b079d0434fe3/src/Features/Core/Portable/Diagnostics/DiagnosticAnalyzerExtensions.cs#L18-L25

Reply:
oh.  will fllowup.